protocols/kad/src/record/store/memory.rs

```diff
@@ -152,38 +152,45 @@ impl RecordStore for MemoryStore {
         }
         .or_insert_with(Default::default);
 
-        if let Some(i) = providers.iter().position(|p| p.provider == record.provider) {
-            // In-place update of an existing provider record.
-            providers.as_mut()[i] = record;
-        } else {
-            // It is a new provider record for that key.
-            let local_key = self.local_key;
-            let key = kbucket::Key::new(record.key.clone());
-            let provider = kbucket::Key::from(record.provider);
-            if let Some(i) = providers.iter().position(|p| {
-                let pk = kbucket::Key::from(p.provider);
-                provider.distance(&key) < pk.distance(&key)
-            }) {
-                // Insert the new provider.
-                if local_key.preimage() == &record.provider {
+        for p in providers.iter_mut() {
+            if p.provider == record.provider {
+                // In-place update of an existing provider record.
+                if self.local_key.preimage() == &record.provider {
+                    self.provided.remove(p);
                     self.provided.insert(record.clone());
                 }
-                providers.insert(i, record);
-                // Remove the excess provider, if any.
-                if providers.len() > self.config.max_providers_per_key {
-                    if let Some(p) = providers.pop() {
-                        self.provided.remove(&p);
-                    }
-                }
-            } else if providers.len() < self.config.max_providers_per_key {
-                // The distance of the new provider to the key is larger than
-                // the distance of any existing provider, but there is still room.
-                if local_key.preimage() == &record.provider {
-                    self.provided.insert(record.clone());
+                *p = record;
+                return Ok(());
+            }
+        }
+
+        // It is a new provider record for that key.
+        let key = kbucket::Key::new(record.key.clone());
+        let provider = kbucket::Key::from(record.provider);
+        let position = providers
+            .iter()
+            .position(|p| key.distance(&provider) <= key.distance(&kbucket::Key::from(p.provider)))
+            .unwrap_or_else(|| providers.len());
```

Comment:
Why do we want to store the provider records of peers whose identity is close to the key they provide in priority?

I see this was already the case in the past, but I think it is wrong. It make sense to have an upper bound in the number of provider records stored for a key, but this number doesn't have to be the magic `K`, it should be customizable. Storing in priority records from peers that are close to the key make censorship of keys easy, since an attacker could forge identities close to the key they want to eclipse, and advertise they have the key (and not serve content upon request). All previous providers will be dropped from the DHT, and the content cannot be discovered.

This can be tackled in a follow up PR.

Reply:
> It make sense to have an upper bound in the number of provider records stored for a key, but this number doesn't have to be the magic K, it should be customizable.

It's already customizable (`max_providers_per_key`)

> Storing in priority records from peers that are close to the key make censorship of keys easy

Sure but please note that storing peers equally has the same problem (but less serious). As long as there's an upper bound in the number of provider records, the attacker can forge enough many identities to eclipse the key. I investigated `go-libp2p-kad-dht` and [it](https://github.com/libp2p/go-libp2p-kad-dht/blob/25d299e64f3cb9a3da337e9d4af1a1b722ae503d/providers/providers_manager.go#L271) seems that it does not impose explicit limit on the number of provider records.

Reply:
Since this upper bound is not mentioned in the [specs](https://github.com/libp2p/specs/tree/master/kad-dht) either, I suggest having no limits in the number of providers per key by default. However if a limit is set and a new provider is added to the store, we can randomly choose a provider record to replace. Does that look good to you?

Reply:
`rust-libp2p` follows the principle to [bound everything](https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#bound-everything), so let's keep a customizable upper bound 👍🏻 

IMO we should get rid of the XOR distance sorting because it isn't appropriate. I can see different solutions to decide which providers to keep:
1. Seniority: if we have enough providers, drop new providers. If an existing provider republishes a record for the same key, update the record. Good for censorship resistance, bad for load balancing.
2. Override: When a new provider is received, override a random existing provider with the new provider. Terrible to censorship resistance, good for load balancing.

I have a preference for the first approach since load balancing should be improved by increasing the `max_providers_per_key`, and there would be no way to compensate for censorship vulnerability. Moreover, the second approach would make the network more random and harder to debug.

Reply:
Sounds good